### PR TITLE
feat(sync): enforce allowed sync folders via platform-specific rules

### PR DESCRIPTION
**Pull request by LawrenceLeitgib**
## Summary
Introduces a new sync-folder rule system that restricts which local directories users may add as sync folders. Platform-specific CSV rule files are loaded into `ParmsDb` on startup and enforced by the server whenever a new sync is created. The Qt GUI now warns the user when an invalid folder is selected.

## Changes
- **New model and persistence:** added `SyncFolderRule` entity and a matching `sync_folder_rule` table in `ParmsDb` with CRUD operations (`src/libparms/db/syncfolderrule.{h,cpp}`).
- **Platform rule files:** added `sync-folder-rules-linux.csv`, `sync-folder-rules-osx.csv`, and `sync-folder-rules-win.csv` to define per-OS allowed and blocked paths.
- **CSV synchronization:** implemented `ParmsDb::getSyncFolderRulesFromFile()` and `updateSyncFolderRules()` to read the platform CSV and keep the DB in sync.
- **Server enforcement:** added `ServerRequests::isSyncFolderAllowedByRules()` which checks candidate paths against the loaded rules before allowing a new sync (`ServerRequests::addSync`, `isPathValidForNewSync`).
- **Environment expansion:** rule paths support `$HOME`, `$USER`, and `~` expansion across all platforms.
- **GUI (legacy) validation:** added `GuiRequests::isPathValidForNewSync()` and `GuiUtility::validateLocalFolderForNewSync()` so the folder-selection dialog rejects disallowed paths with a warning message.
- **Build and packaging:** updated CMake, GitHub Actions, MSI WiX, Linux package scripts, and AppImage configurations to bundle the new CSV files.
- **Tests:** added `test/server/requests/testserverrequests` covering rule matching logic, environment-variable expansion, and edge cases.

## Technical Details
Rules are evaluated using a "deepest-match-wins" strategy. Each rule is expanded and compared against the candidate directory. The deepest matching rule type decides the outcome:
 
* `BlackList` - path is rejected.
* `WhiteList` - path is accepted.
* `WhiteListSubFolder` - acceptance is granted only if the candidate is a strict subfolder of the rule path.
If no rule matches at all, the path is accepted by default (fail-open). This ensures that removable media (such as a USB key mounted on its own drive or mount point) remains syncable without requiring an explicit rule for every possible location.

## Testing
- The new rule matching logic is covered by the `TestServerRequests` suite.

### CODE REVIEW

File: .github/actions/build_macos/action.yml [hunk 1]
```diff
@@ -131,8 +131,10 @@ runs:
           mkdir uncached_dependencies
           if [ "${{ inputs.release_build }}" = "true" ]; then
                 cp ${{ github.workspace }}/build-macos/client/install/kDrive.app/Contents/Resources/*.lst ./uncached_dependencies/
+                cp ${{ github.workspace }}/build-macos/client/install/kDrive.app/Contents/Resources/sync-folder-rules*.csv ./uncached_dependencies/
           else
                 cp ${{ github.workspace }}/build-macos/client/bin/*.lst ./uncached_dependencies/
+                cp ${{ github.workspace }}/build-macos/client/bin/sync-folder-rules*.csv ./uncached_dependencies/
```
Comment (ClementKunz) [inline .github/actions/build_macos/action.yml]: Why `sync-folder-rules*.csv` and not just `sync-folder-rules.csv`?

File: src/libcommonserver/utility/utility.cpp [hunk 2]
```diff
@@ -435,6 +436,13 @@ bool Utility::normalizedSyncPath(const SyncPath &path, SyncPath &normalizedPath,
 
     return true;
 }
+SyncPath Utility::getSyncFolderRulesFilePath(const bool test) {
+    if (test) return syncFolderRulesFileName;
```
Comment (ClementKunz) [inline src/libcommonserver/utility/utility.cpp:440]: Avoid adding some logic only for testing. Prefer mocked objects in tests when possible.
This method seems to be called only in parmsDB.cpp so maybe you should move this implementation to `ParmsDB`, which would simplify mocking.

File: src/libparms/db/parmsdb.cpp [hunk 3]
```diff
@@ -837,6 +857,140 @@ bool ParmsDb::insertUserTemplateNormalizations(const std::string &fromVersion) {
     return result;
 }
 
+
+namespace {
+
+std::string trim(const std::string_view str) {
```
Comment (ClementKunz) [inline src/libparms/db/parmsdb.cpp]: A `trim` method already exists in `CommonUtility`

File: src/libparms/db/parmsdb.cpp [hunk 4]
```diff
@@ -837,6 +857,140 @@ bool ParmsDb::insertUserTemplateNormalizations(const std::string &fromVersion) {
...
+    const size_t start = str.find_first_not_of(' ');
+    if (start == std::string_view::npos) return "";
+    const size_t end = str.find_last_not_of(' ');
+    return std::string(str.substr(start, end - start + 1));
+}
+void stripLineEndings(std::string &line) {
```
Comment (ClementKunz) [inline src/libparms/db/parmsdb.cpp:869]: So maybe we could move this in `CommonUtility` with other string utility functions

File: src/libparms/db/parmsdb.cpp [hunk 5]
```diff
@@ -837,6 +857,140 @@ bool ParmsDb::insertUserTemplateNormalizations(const std::string &fromVersion) {
...
+bool ParmsDb::updateSyncFolderRules() {
+    std::vector<SyncFolderRule> dbSyncFolderRules;
+    if (!selectAllSyncFolderRules(dbSyncFolderRules)) {
+        LOG_WARN(_logger, "Error in selectAllSyncFolderRules");
+        return false;
+    }
+
+    std::vector<SyncFolderRule> fileSyncFolderRules;
+    if (const auto &syncFolderRulesFileName = Utility::getSyncFolderRulesFilePath(_test);
+        !getSyncFolderRulesFromFile(syncFolderRulesFileName.c_str(), fileSyncFolderRules)) {
```
Comment (ClementKunz) [inline src/libparms/db/parmsdb.cpp]: Why do you convert `syncFolderRulesFileName` to char*? `getSyncFolderRulesFilePath` returns a `SyncPath` and `getSyncFolderRulesFromFile` takes a `SyncPath` as parameter.
Also, maybe rename `syncFolderRulesFileName`. It contains a full file path, not just the name

File: src/libparms/db/parmsdb.cpp [hunk 6]
```diff
@@ -2938,6 +3119,96 @@ bool ParmsDb::updateAllExclusionApps(const bool def, const std::vector<Exclusion
...
+    return true;
+}
+
+bool ParmsDb::selectAllSyncFolderRules(std::vector<SyncFolderRule> &syncFolderRuleList) {
+    const std::scoped_lock lock(_mutex);
+
+    syncFolderRuleList.clear();
+
+    LOG_IF_FAIL(queryResetAndClearBindings(SELECT_ALL_SYNC_FOLDER_RULE_REQUEST_ID));
+    bool found;
```
Comment (ClementKunz) [inline src/libparms/db/parmsdb.cpp]: Always initialize your variable

File: src/libparms/db/syncfolderrule.cpp [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "syncfolderrule.h"
+
+namespace KDC {
+SyncFolderRule::SyncFolderRule() :
+    _folderRuleType(SyncFolderRuleType::BlackList) {}
```
Comment (ClementKunz) [inline src/libparms/db/syncfolderrule.cpp]: We usually prefer to initialize variables with default values in the header file (`SyncFolderRuleType _folderRuleType{SyncFolderRuleType::BlackList};`). Moreover, this would allow you to remove this implementation of a default constructor and simply use the `default` keyword in the header file.
This is a nitpick, but it is, IMO, better for readability.

File: src/server/requests/serverrequests.cpp [hunk 8]
```diff
@@ -1150,6 +1174,123 @@ ExitCode ServerRequests::createSync(const Sync &sync, SyncInfo &syncInfo) {
     return ExitCode::Ok;
 }
 
+ExitInfo ServerRequests::isSyncFolderAllowedByRules(const SyncPath &path, bool &allowed) {
```
Comment (ClementKunz) [inline src/server/requests/serverrequests.cpp]: This is a big function, and `ServerRequests` is already huge. Maybe this logic should be moved into a dedicated class.

File: src/server/requests/serverrequests.cpp [hunk 9]
```diff
@@ -1150,6 +1174,123 @@ ExitCode ServerRequests::createSync(const Sync &sync, SyncInfo &syncInfo) {
...
+        LOG_WARN(Log::instance()->getLogger(), "Error in ParmsDb::selectAllSyncFolderRules");
+        return ExitCode::DbError;
+    }
+
+    if (rules.empty()) {
+        // Technically we only allow a directory that is whitelisted, but it's assumed that if something is wrong with the rules
+        // file we don't just reject every directory
+        LOG_DEBUG(Log::instance()->getLogger(), "isSyncFolderAllowedByRules: no rules found, allowing path");
+        return ExitCode::Ok;
+    }
```
Comment (ClementKunz) [inline src/server/requests/serverrequests.cpp]: Ok for the fallback, but this should never happen. Maybe send a sentry message if this occurs.

File: src/server/requests/serverrequests.cpp [hunk 10]
```diff
@@ -1150,6 +1174,123 @@ ExitCode ServerRequests::createSync(const Sync &sync, SyncInfo &syncInfo) {
...
+
+    LOG_DEBUG(Log::instance()->getLogger(), "isSyncFolderAllowedByRules: found " << rules.size() << " rules");
+
+    const QString candidateDir = QDir::cleanPath(Path2QStr(path)) + '/';
+
+    const SyncFolderRule *bestMatch = nullptr;
+    SyncPath bestMatchExpandedPath;
+    std::int32_t bestDepth = -1;
```
Comment (ClementKunz) [inline src/server/requests/serverrequests.cpp]:
```suggestion
    int32_t bestDepth = -1;
```

File: src/server/requests/serverrequests.cpp [hunk 11]
```diff
@@ -1150,6 +1174,123 @@ ExitCode ServerRequests::createSync(const Sync &sync, SyncInfo &syncInfo) {
...
+            (void) pathStr.replace("$USER", user);
+        }
+
+        // Handle ~ shorthand
+        if (pathStr.startsWith("~")) {
+            (void) pathStr.replace(0, 1, homeDir);
+        }
+
+        return QStr2Path(pathStr);
+    };
```
Comment (ClementKunz) [inline src/server/requests/serverrequests.cpp]: We are currently trying to move away from Qt on the server side. Therefore, you should avoid using it here.
I have a PR waiting to be merged with a new utility function to retrieved the HOMe folder: https://github.com/Infomaniak/desktop-kDrive/pull/2050/changes#diff-68322adf7cccc8797463fd7b96f6099531fc0e22593c456e5d91d524e1b7acf9

File: .github/actions/build_macos/action.yml (hunk 1 above)
Comment (LawrenceLeitgib) [reply .github/actions/build_macos/action.yml]: https://github.com/Infomaniak/desktop-kDrive/pull/2124/commits/557ac72d14cea15a33460cf668c1a7247ca55d1d

File: src/libcommonserver/utility/utility.cpp (hunk 2 above)
Comment (LawrenceLeitgib) [reply src/libcommonserver/utility/utility.cpp:440]: It's actually needed to pass the MacOS CI

File: src/libparms/db/parmsdb.cpp (hunk 3 above)
Comment (LawrenceLeitgib) [reply src/libparms/db/parmsdb.cpp]: https://github.com/Infomaniak/desktop-kDrive/pull/2124/commits/68acd1764811c8e8c9694f5f4e86a2dcef4d6390

File: src/libparms/db/parmsdb.cpp (hunk 4 above)
Comment (LawrenceLeitgib) [reply src/libparms/db/parmsdb.cpp:869]: I choose to keep it there for now as I think this function is very specific to this task, to discuss though

File: src/libparms/db/parmsdb.cpp (hunk 5 above)
Comment (LawrenceLeitgib) [reply src/libparms/db/parmsdb.cpp]: https://github.com/Infomaniak/desktop-kDrive/pull/2124/commits/7669fb6d95a3612e977f9841216ce76aeb175b65

File: src/libparms/db/parmsdb.cpp (hunk 6 above)
Comment (LawrenceLeitgib) [reply src/libparms/db/parmsdb.cpp]: https://github.com/Infomaniak/desktop-kDrive/pull/2124/commits/3fb598c0829637b547acdedcd7124308d2236c8b

File: src/server/requests/serverrequests.cpp (hunk 9 above)
Comment (LawrenceLeitgib) [reply src/server/requests/serverrequests.cpp]: https://github.com/Infomaniak/desktop-kDrive/pull/2124/commits/ac3611e52acb8715f7683ff2cdfe4c2c7d0631bb

File: src/server/requests/serverrequests.cpp (hunk 10 above)
Comment (LawrenceLeitgib) [reply src/server/requests/serverrequests.cpp]: https://github.com/Infomaniak/desktop-kDrive/pull/2124/commits/2b571627f92c375df77c960b594ced4a2962effc

File: src/libparms/db/syncfolderrule.cpp (hunk 7 above)
Comment (LawrenceLeitgib) [reply src/libparms/db/syncfolderrule.cpp]: https://github.com/Infomaniak/desktop-kDrive/pull/2124/commits/6a465dd2d221810fdd0eb21676e36acff559d9fd

File: src/server/requests/serverrequests.cpp (hunk 8 above)
Comment (LawrenceLeitgib) [reply src/server/requests/serverrequests.cpp]: https://github.com/Infomaniak/desktop-kDrive/pull/2124/commits/65c5d051f4584229d5792f49a525ccf8bfa55aea

File: test/server/requests/testserverrequests.cpp [hunk 12]
```diff
@@ -312,5 +313,129 @@ void TestServerRequests::testFolderContainsNonExcludedItemMixed() {
...
+    clearRules();
+    insertRule("/home/user", SyncFolderRuleType::WhiteListSubFolder);
+    insertRule("/home/user/.local/share/Trash", SyncFolderRuleType::BlackList);
+    bool allowed = false;
+    CPPUNIT_ASSERT_EQUAL(ExitInfo(ExitCode::Ok), SyncFolderAllowedChecker::check("/home/user/Documents", allowed));
+    CPPUNIT_ASSERT(allowed);
+
+    allowed = true;
+    CPPUNIT_ASSERT_EQUAL(ExitInfo(ExitCode::Ok),
+                         SyncFolderAllowedChecker::check("/home/user/.local/share/Trash/file.txt", allowed));
```
Comment (ClementKunz) [inline test/server/requests/testserverrequests.cpp]: A bit strange to provide a file here instead of a path. I know that we check if the provided path points toward a folder outside the `SyncFolderAllowedChecker::check` so it is a valid test. But for the sake of readability and simplicity, perhaps we can just pass folders in those tests.

File: sync-folder-rules-win.csv [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+$HOME,WhiteListSubFolder
+$HOME/AppData,BlackList
+$HOME/NTUSER.DAT,BlackList
+$HOME/ntuser.dat.LOG1,BlackList
+$HOME/ntuser.dat.LOG2,BlackList
+$HOME/Application Data,BlackList
+$HOME/$RECYCLE.BIN,BlackList
+D:\,WhiteListSubFolder
+D:\$RECYCLE.BIN,BlackList
+D:\System Volume Information,BlackList
```
Comment (ClementKunz) [inline sync-folder-rules-win.csv]: On Windows, you can have any letter to describe your disc. So hardcodding "D:" here is not enough

File: test/server/requests/testserverrequests.cpp
Comment (ClementKunz) [inline test/server/requests/testserverrequests.cpp:1]: Perhaps you can add a few tests more OS specific. For example testing with Windows specific path and different disc root ("C:", "D:", ...)

File: test/server/requests/testserverrequests.cpp (hunk 12 above)
Comment (LawrenceLeitgib) [reply test/server/requests/testserverrequests.cpp]: https://github.com/Infomaniak/desktop-kDrive/pull/2124/commits/2298e393c2b58af5dbebbd4b0eefae5b0fe0f2fd

File: test/server/requests/testserverrequests.cpp
Comment (LawrenceLeitgib) [reply test/server/requests/testserverrequests.cpp:1]: [b341070](https://github.com/Infomaniak/desktop-kDrive/pull/2124/commits/b341070320707bc24d4f7491bd558f623b624792)

File: sync-folder-rules-win.csv (hunk 13 above)
Comment (LawrenceLeitgib) [reply sync-folder-rules-win.csv]: [b341070](https://github.com/Infomaniak/desktop-kDrive/pull/2124/commits/b341070320707bc24d4f7491bd558f623b624792)

File: src/server/requests/serverrequests.cpp (hunk 11 above)
Comment (LawrenceLeitgib) [reply src/server/requests/serverrequests.cpp]: https://github.com/Infomaniak/desktop-kDrive/pull/2124/commits/b341070320707bc24d4f7491bd558f623b624792

Comment (LawrenceLeitgib) [reply src/server/requests/serverrequests.cpp]: et ce commit enlève complétement QT de la function [30657db](https://github.com/Infomaniak/desktop-kDrive/pull/2124/commits/30657db13bc8fd0c956cb66172bef3060bc76b8d)

File: src/libparms/db/parmsdb.cpp [hunk 14]
```diff
@@ -843,6 +863,134 @@ bool ParmsDb::insertUserTemplateNormalizations(const std::string &fromVersion) {
...
+        }
+
+        if (!headerSkipped) {
+            headerSkipped = true;
+            if (line.find("Path") != std::string::npos) {
+                continue;
+            }
+        }
+
+        tryParseCsvLine(line, fileSyncFolderRules, _logger);
```
Comment (ClementKunz) [inline src/libparms/db/parmsdb.cpp]: You don't use the return value here. Since this method is called only here, either use the return value or make tryParseCsvLine return `void`.

File: src/server/requests/syncfolderallowedchecker.cpp [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+    if (rules.empty()) {
+        LOG_DEBUG(Log::instance()->getLogger(), "isSyncFolderAllowedByRules: no rules found, allowing path");
+        sentry::Handler::captureMessage(sentry::Level::Warning, "ServerRequests::isSyncFolderAllowedByRules",
+                                        "No sync rules found, allowing path");
+        return ExitCode::Ok;
+    }
+
+    LOG_DEBUG(Log::instance()->getLogger(), "isSyncFolderAllowedByRules: found " << rules.size() << " rules");
+
+    QString candidateDir = QDir::cleanPath(Path2QStr(path));
```
Comment (ClementKunz) [inline src/server/requests/syncfolderallowedchecker.cpp]: Can we avoid Qt here?

File: src/server/requests/syncfolderallowedchecker.cpp [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+
+        SyncPath expandedRulePath;
+        if (const auto exitInfo = expandRulePath(rule.syncPath(), expandedRulePath); !exitInfo) {
+            return exitInfo;
+        }
+
+        LOGW_DEBUG(Log::instance()->getLogger(),
+                   L"isSyncFolderAllowedByRules: expanded rule syncPath: " << Utility::formatSyncPath(expandedRulePath));
+
+        QString ruleDir = QDir::cleanPath(Path2QStr(expandedRulePath));
```
Comment (ClementKunz) [inline src/server/requests/syncfolderallowedchecker.cpp]: Same, no Qt if possible

File: test/server/requests/testserverrequests.cpp
Comment (ClementKunz) [inline test/server/requests/testserverrequests.cpp:1]: Maybe just add some more OS specific test. for exemple:
- on `E:` for Windows
- on `/Volume` for macOS
- on `/media` for Linux

File: src/libparms/db/parmsdb.cpp (hunk 14 above)
Comment (LawrenceLeitgib) [reply src/libparms/db/parmsdb.cpp]: I made it log a warning because nothing should be fatal in case of failing to parse a line in the csv [5cb3c52](https://github.com/Infomaniak/desktop-kDrive/pull/2124/commits/5cb3c525fa67d21e872a27b41978d81ece634477)

File: src/server/requests/syncfolderallowedchecker.cpp (hunk 15 above)
Comment (LawrenceLeitgib) [reply src/server/requests/syncfolderallowedchecker.cpp]: Yes, doable https://github.com/Infomaniak/desktop-kDrive/pull/2124/commits/8f336d90dd9c5250d4ee3245957d1c8f08658b0a

File: src/server/requests/syncfolderallowedchecker.cpp (hunk 16 above)
Comment (LawrenceLeitgib) [reply src/server/requests/syncfolderallowedchecker.cpp]: same as above https://github.com/Infomaniak/desktop-kDrive/pull/2124/commits/8f336d90dd9c5250d4ee3245957d1c8f08658b0a

File: test/server/requests/testserverrequests.cpp
Comment (LawrenceLeitgib) [reply test/server/requests/testserverrequests.cpp:1]: Here are the OS specific test https://github.com/Infomaniak/desktop-kDrive/pull/2124/commits/945b9ddeacdaedcd683423048041ac01d0d16abd

**Review by ClementKunz — APPROVED** [review]
